[L0] Refactoring of boolean event parameters

winstonzhang-intel opened the pull request:
CI in LLVM/SYCL: https://github.com/intel/llvm/pull/14536

nrspruit commented on source/adapters/level_zero/common.hpp:
```diff
@@ -241,19 +248,6 @@ static const uint32_t UrL0QueueSyncNonBlocking = [] {
   return L0QueueSyncLockingModeValue;
 }();
 
-// Controls whether the L0 Adapter creates signal events for commands on
```
why was this deleted? You have not changed the usage of that env and this is needed. Please keep this.

nrspruit commented on source/adapters/level_zero/event.cpp:
```diff
@@ -1247,16 +1253,15 @@ ur_result_t EventCreate(ur_context_handle_t Context, ur_queue_handle_t Queue,
   ZeEventDesc.wait = 0;
 
-  if (HostVisible || CounterBasedEventEnabled) {
+  if (Flags & USING_IMM_CMDLIST || Flags & COUNTER_BASED) {
```
```suggestion
  if (Flags & HOST_VISIBLE || Flags & COUNTER_BASED) {
```

nrspruit reviewed (changes requested):
Looks good,  I found one typo in the flag checks, please fix that and verify that there are not others I missed, thanks!

pbalcer commented on source/adapters/level_zero/context.hpp:
```diff
@@ -222,45 +217,42 @@ struct ur_context_handle_t_ : _ur_object {
-                                  bool UsingImmediateCmdList,
+  ur_result_t calculateCacheIndex(EventFlags Flags,
                                   EventPoolCacheType &CacheType) {
-    if (CounterBasedEventEnabled && HostVisible && !UsingImmediateCmdList) {
```
hm, can't we just use Flags as the index?

winstonzhang-intel replied:
As in using the flags for the index of zeEventPoolCache?

pbalcer replied:
Yes. Notice that the number of combinations in your cache grows exponentially with the number of boolean flags (+ potentially number of bits you'd need to dedicate to store the "id" of a device). So a flag, with the corresponding bits set, is your index into the cache.

pbalcer commented on source/adapters/level_zero/common.hpp:
```diff
@@ -218,6 +218,13 @@ enum {
+  HOST_VISIBLE = 1 << 2,
+  ENABLE_PROFILER = 1 << 3,
+  MULTIDEVICE = 1 << 4
+} EventFlags;
```
Storing invalid sequences in an enum is UB. You should have uint16_t or something to store the actual bit values, like in the UR API:
https://github.com/oneapi-src/unified-runtime/blob/main/include/ur_api.h#L552

winstonzhang-intel replied:
Done

pbalcer commented on source/adapters/level_zero/context.cpp:
```diff
@@ -580,8 +578,11 @@ void ur_context_handle_t_::addEventToContextCache(ur_event_handle_t Event) {
-  auto Cache = getEventCache(Event->isHostVisible(),
-                             Event->isProfilingEnabled(), Device);
+  EventFlags Flags = static_cast<EventFlags>(
+      Event->isHostVisible()            ? HOST_VISIBLE
```
maybe we can just store flags in the Event?

winstonzhang-intel replied:
At least for isHostVisible() it is stored as a pointer to itself if it is hostvisible, which other events will use. We could add a ur_event_flag for isHostVisible but wouldn't see any performance gain since isHostVisible is const.

pbalcer replied:
I just thought it might make the code simpler. In the places where currently you need to calculate the flags for an event to find the cache, you'd be able to simply do something like `getCache(event->flags, event->device)` (simplified).

nrspruit commented:
Hello @winstonzhang-intel ,

Please rebase since there are L0 CTS failures here: https://github.com/oneapi-src/unified-runtime/actions/runs/9651142091/job/26618341665?pr=1600.

nrspruit reviewed (approved):
LGTM

omarahmed1111 commented:
@winstonzhang-intel Please make sure that https://github.com/intel/llvm/pull/14536 is updated and passing CI, so we could merge this. Thanks!

omarahmed1111 commented:
I removed the ready to merge label temporarily until the intel/llvm side is updated and passing tests. Please readd it when intel/llvm PR is updated.

winstonzhang-intel commented:
> I removed the ready to merge label temporarily until the intel/llvm side is updated and passing tests. Please readd it when intel/llvm PR is updated.



> I removed the ready to merge label temporarily until the intel/llvm side is updated and passing tests. Please readd it when intel/llvm PR is updated.

Working on rebasing against latest phase 2 of counterbased patch.

nrspruit commented:
@winstonzhang-intel , please resolve the merge conflicts with a rebase otherwise the changes look good.

kbenzie commented:
@winstonzhang-intel unfortunately I had to revert this due to the CI failures on https://github.com/intel/llvm/pull/14536, they are very confusing to me.

kbenzie commented:
This could be related to https://github.com/intel/llvm/pull/16747#issuecomment-2652441043 which I think should have been resolved by https://github.com/oneapi-src/unified-runtime/pull/2690 but will need to investigate further.

winstonzhang-intel commented:
> This could be related to [intel/llvm#16747 (comment)](https://github.com/intel/llvm/pull/16747#issuecomment-2652441043) which I think should have been resolved by #2690 but will need to investigate further.

Sounds good, please let me know. Thanks.
